internal/download/downloader/downloader_test.go

```diff
@@ -0,0 +1,53 @@
+package downloader
+
+import (
+	"context"
+	"errors"
+	"testing"
+	"time"
+
+	"go.sia.tech/renterd/api"
+	"go.sia.tech/renterd/internal/test/mocks"
+	"go.sia.tech/renterd/internal/utils"
+)
+
+func TestDownloaderStopped(t *testing.T) {
+	assertErr := func(t *testing.T, req SectorDownloadReq, expected error) {
+		select {
+		case <-req.Resps.Received():
+			if err := req.Resps.responses[0].Err; !utils.IsErr(err, expected) {
+				t.Fatal("unexpected error response", err)
+			}
+		case <-time.After(time.Second):
+			t.Fatal("no response")
+		}
+	}
+
+	t.Run("stop before enqueue", func(t *testing.T) {
+		hm := mocks.NewHostManager()
+		dl := New(context.Background(), hm.Downloader(api.HostInfo{}))
+		req := SectorDownloadReq{
+			Ctx:   context.Background(),
+			Resps: NewSectorResponses(),
+		}
+
+		dl.Stop(errors.New(t.Name()))
+		dl.Enqueue(&req)
+
+		assertErr(t, req, ErrStopped)
+	})
+
+	t.Run("stop after enqueue", func(t *testing.T) {
+		hm := mocks.NewHostManager()
+		dl := New(context.Background(), hm.Downloader(api.HostInfo{}))
+		req := SectorDownloadReq{
+			Ctx:   context.Background(),
+			Resps: NewSectorResponses(),
+		}
+
+		dl.Enqueue(&req)
+		dl.Stop(errors.New(t.Name()))
```

Comment:
I'm surprised this has never caused NDFs considering the download might be done before `Stop` is called.

Reply:
The downloader is not processing its queue.